File: src/backendng/src/main/kotlin/com/secman/service/AppSettingsService.kt
package com.secman.service

import com.secman.domain.AppSettings
import com.secman.repository.AppSettingsRepository
import com.secman.config.AiRiskAssessmentConfig
import io.micronaut.context.annotation.Value
import io.micronaut.serde.annotation.Serdeable
import jakarta.inject.Singleton
import jakarta.transaction.Transactional
import org.slf4j.LoggerFactory

/**
 * Service for managing application-wide settings.
 * Feature: 068-requirements-alignment-process
 */
@Singleton
open class AppSettingsService(
    private val appSettingsRepository: AppSettingsRepository,
    private val aiRiskAssessmentConfig: AiRiskAssessmentConfig
) {
    private val logger = LoggerFactory.getLogger(AppSettingsService::class.java)

    @Value("\${app.frontend.base-url:http://localhost:4321}")
    private var defaultBaseUrl: String = "http://localhost:4321"

    /**
     * Feature 088 seed value. Used only when the app_settings row is created
     * for the first time. After that, the DB column is authoritative — the
     * env var being flipped will not change runtime behaviour.
     */
    @Value("\${secman.ai.risk-assessment.enabled:false}")
    private var aiRiskAssessmentDefault: Boolean = false

    @Serdeable
    data class AppSettingsDto(
        val baseUrl: String,
        val globalCveApprovalAdminOnly: Boolean,
        val aiRiskAssessmentEnabled: Boolean,
        val aiRiskAssessmentModel: String,
        val updatedBy: String?,
        val updatedAt: String?
    )

    /**
     * Get current application settings.
     * Creates default settings if none exist.
     */
    @Transactional
    open fun getSettings(): AppSettingsDto {
        val settings = getOrCreateSettings()
        return AppSettingsDto(
            baseUrl = settings.getNormalizedBaseUrl(),
            globalCveApprovalAdminOnly = settings.globalCveApprovalAdminOnly,
            aiRiskAssessmentEnabled = settings.aiRiskAssessmentEnabled,
            aiRiskAssessmentModel = settings.aiRiskAssessmentModel,
            updatedBy = settings.updatedBy,
            updatedAt = settings.updatedAt?.toString()
        )
    }

    /**
     * Effective feature flag for the AI risk-assessment feature. Source of
     * truth for `ComplianceAssistantService.isEnabled()` — note this also
     * requires an API key to be configured separately.
     */
    @Transactional
    open fun isAiRiskAssessmentEnabled(): Boolean {
        return getOrCreateSettings().aiRiskAssessmentEnabled
    }

    @Transactional
    open fun getAiRiskAssessmentModel(): String {
        return getOrCreateSettings().aiRiskAssessmentModel
    }

    /**
     * Get the base URL for use in email generation.
     * Returns the normalized URL without trailing slash.
     */
    @Transactional
    open fun getBaseUrl(): String {
        val settings = getOrCreateSettings()
        return settings.getNormalizedBaseUrl()
    }

    /**
     * Check if global CVE approval is restricted to ADMIN only.
     */
    @Transactional
    open fun isGlobalCveApprovalAdminOnly(): Boolean {
        return getOrCreateSettings().globalCveApprovalAdminOnly
    }

    /**
     * Update application settings.
     *
     * @param baseUrl The new base URL
     * @param updatedBy Username of the admin making the change
     * @param globalCveApprovalAdminOnly Whether CVE_PATTERN approvals require ADMIN role
     * @return Updated settings DTO
     * @throws IllegalArgumentException if validation fails
     */
    @Transactional
    open fun updateSettings(
        baseUrl: String,
        updatedBy: String,
        globalCveApprovalAdminOnly: Boolean = false,
        aiRiskAssessmentEnabled: Boolean = false,
        aiRiskAssessmentModel: String = aiRiskAssessmentConfig.model
    ): AppSettingsDto {
        val settings = getOrCreateSettings()

        // Update the base URL
        settings.baseUrl = baseUrl.trimEnd('/')
        settings.globalCveApprovalAdminOnly = globalCveApprovalAdminOnly
        settings.aiRiskAssessmentEnabled = aiRiskAssessmentEnabled
        settings.aiRiskAssessmentModel = aiRiskAssessmentModel.trim()
        settings.updatedBy = updatedBy

        if (settings.aiRiskAssessmentModel.isBlank()) {
            throw IllegalArgumentException("AI risk-assessment model cannot be empty")
        }

        // Validate
        val errors = settings.validateBaseUrl()
        if (errors.isNotEmpty()) {
            throw IllegalArgumentException(errors.joinToString("; "))
        }

        val updated = appSettingsRepository.update(settings)
        logger.info(
            "App settings updated by {}: baseUrl={}, globalCveApprovalAdminOnly={}, aiRiskAssessmentEnabled={}",
            updatedBy, updated.baseUrl, updated.globalCveApprovalAdminOnly, updated.aiRiskAssessmentEnabled
        )

        return AppSettingsDto(
            baseUrl = updated.getNormalizedBaseUrl(),
            globalCveApprovalAdminOnly = updated.globalCveApprovalAdminOnly,
            aiRiskAssessmentEnabled = updated.aiRiskAssessmentEnabled,
            aiRiskAssessmentModel = updated.aiRiskAssessmentModel,
            updatedBy = updated.updatedBy,
            updatedAt = updated.updatedAt?.toString()
        )
    }

    /**
     * Get existing settings or create default ones.
     */
    private fun getOrCreateSettings(): AppSettings {
        val existing = appSettingsRepository.findAll()
        return if (existing.isEmpty()) {
            logger.info(
                "Creating default app settings: baseUrl={}, aiRiskAssessmentEnabled={} (from env default)",
                defaultBaseUrl, aiRiskAssessmentDefault
            )
            val default = AppSettings(
                baseUrl = defaultBaseUrl,
                aiRiskAssessmentEnabled = aiRiskAssessmentDefault,
                aiRiskAssessmentModel = aiRiskAssessmentConfig.model,
                updatedBy = "system"
            )
            appSettingsRepository.save(default)
        } else {
            existing.first()
        }
    }
}

File: src/frontend/src/services/appSettingsService.ts
import { authenticatedGet, authenticatedPut } from '../utils/auth';

/**
 * Service for Application Settings API operations
 * Feature: 068-requirements-alignment-process
 *
 * Provides methods to get and update application-wide settings
 * such as the base URL used in email notifications.
 */

/**
 * Application settings configuration
 */
export interface AppSettingsDto {
  /** Base URL of the application frontend (e.g., https://secman.example.com) */
  baseUrl: string;

  /** Username of admin who last updated settings */
  updatedBy: string | null;

  /** Timestamp of last update */
  updatedAt: string | null;
}

/**
 * Standard error response
 */
export interface ErrorResponse {
  status: number;
  message: string;
  timestamp: string;
}

/**
 * Get current application settings
 *
 * Requires: ADMIN role
 *
 * @returns Current application settings
 * @throws Error if request fails or user lacks ADMIN role
 */
export async function getAppSettings(): Promise<AppSettingsDto> {
  console.log('[appSettingsService] getAppSettings called');
  console.log('[appSettingsService] Making authenticated GET to /api/settings/app');

  const response = await authenticatedGet('/api/settings/app');

  console.log('[appSettingsService] Response received:', {
    ok: response.ok,
    status: response.status,
    statusText: response.statusText
  });

  if (!response.ok) {
    console.error('[appSettingsService] Request failed with status:', response.status);
    const errorData = await response.json().catch(() => ({ message: 'Request failed' }));
    console.error('[appSettingsService] Error data:', errorData);
    throw new Error(errorData.message || `Request failed with status ${response.status}`);
  }

  const data = await response.json();
  console.log('[appSettingsService] Successfully parsed response:', data);
  return data;
}

/**
 * Update application settings
 *
 * Requires: ADMIN role
 *
 * @param baseUrl New base URL
 * @returns Updated application settings
 * @throws Error if request fails, validation fails, or user lacks ADMIN role
 */
export async function updateAppSettings(baseUrl: string): Promise<AppSettingsDto> {
  console.log('[appSettingsService] updateAppSettings called with:', baseUrl);
  console.log('[appSettingsService] Making authenticated PUT to /api/settings/app');

  const response = await authenticatedPut('/api/settings/app', { baseUrl });

  console.log('[appSettingsService] Response received:', {
    ok: response.ok,
    status: response.status,
    statusText: response.statusText
  });

  if (!response.ok) {
    console.error('[appSettingsService] Update failed with status:', response.status);
    const errorData = await response.json().catch(() => ({ message: 'Update failed' }));
    console.error('[appSettingsService] Error data:', errorData);
    throw new Error(errorData.message || `Update failed with status ${response.status}`);
  }

  const data = await response.json();
  console.log('[appSettingsService] Successfully updated settings:', data);
  return data;
}
